rsconnect/api.py

```diff
@@ -42,21 +43,63 @@ def handle_bad_response(self, response):
             # search page so trap that since we know we're expecting JSON from Connect.  This
             # also catches all error conditions which we will report as "not running Connect".
             else:
-                if response.json_data and "error" in response.json_data:
-                    error = "The Connect server reported an error: %s" % response.json_data["error"]
+                if response.json_data and "error" in response.json_data and response.json_data["error"] is not None:
+                    error = "%s reported an error (calling %s): %s" % (
+                        self.remote_name,
+                        response.full_uri,
+                        response.json_data["error"],
+                    )
                     raise RSConnectException(error)
                 if response.status < 200 or response.status > 299:
                     raise RSConnectException(
-                        "Received an unexpected response from RStudio Connect: %s %s"
-                        % (response.status, response.reason)
+                        "Received an unexpected response from %s (calling %s): %s %s"
+                        % (self.remote_name, response.full_uri, response.status, response.reason)
                     )
 
 
-class RSConnect(HTTPServer):
-    def __init__(self, server, cookies=None, timeout=30):
+class ShinyappsServer(AbstractRemoteServer):
+    """
+    A simple class to encapsulate the information needed to interact with an
+    instance of the shinyapps.io server.
+    """
+
+    def __init__(self, url: str, account_name: str, token: str, secret: str):
+        super().__init__(url, "shinyapps.io")
+        self.account_name = account_name
+        self.token = token
+        self.secret = secret
+
+
+class RSConnectServer(AbstractRemoteServer):
+    """
+    A simple class to encapsulate the information needed to interact with an
+    instance of the Connect server.
+    """
+
+    def __init__(self, url, api_key, insecure=False, ca_data=None):
+        super().__init__(url, "RStudio Connect")
+        self.api_key = api_key
+        self.insecure = insecure
+        self.ca_data = ca_data
+        # This is specifically not None.
+        self.cookie_jar = CookieJar()
+
+
+TargetableServer = typing.Union[ShinyappsServer, RSConnectServer]
+
+
+class S3Server(AbstractRemoteServer):
+    remote_name = "S3"
+
+    def __init__(self, url: str):
+        self.url = url
```

Comment:
Is there a reason `remote_name` is handled differently in this case compared to the above two. I would have expected...

```
    def __init__(self, url):
        super().__init__(url, "S3")
```

Reply:
Nope, that was just an oversight.